lib/async_hooks.js

```diff
@@ -335,19 +335,14 @@ function emitInitS(asyncId, type, triggerAsyncId, resource) {
     throw new RangeError('triggerAsyncId must be an unsigned integer');
 
   init(asyncId, type, triggerAsyncId, resource);
-
-  // Isn't null if hooks were added/removed while the hooks were running.
-  if (tmp_active_hooks_array !== null) {
-    restoreTmpHooks();
-  }
 }
 
 function emitHookFactory(symbol, name) {
   // Called from native. The asyncId stack handling is taken care of there
   // before this is called.
   // eslint-disable-next-line func-style
   const fn = function(asyncId) {
-    processing_hook = true;
+    processing_hook += 1;
     // Use a single try/catch for all hook to avoid setting up one per
     // iteration.
     try {
```

Comment:
> I'm worried that doing reference counting is too fragile.

Could you be more explicit? 

> Why not replace the for either with a `.forEach`

`.forEach` is not resistant to mutation.

```js
const a = [1,2,3];
a.forEach(function (value) {
  process._rawDebug(value);
  a.pop();
});
```
prints `1 2`.

> with a `for (const hook of Array.from(active_hooks_array))`

Array copy is too slow. When `async_hooks` is enabled this is the hottest code path we have.